File: 78-Subsets-子集/78-Subsets-子集.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
 let res = []
 let path = []  
 var subsets = function(nums) {
     res = []
     backtrace(nums, 0)
     return res
 };  
 
 var backtrace = function(nums, index){
     if(index > nums.length) return
 
     res.push(Array.from(path))
 
     for(let i = index; i < nums.length; i++){
         path.push(nums[i])
         backtrace(nums, i + 1)
         path.pop()
     }
 }

File: 78-Subsets-子集/78-Subsets-子集.java
class Solution {
    public List<List<Integer>> res;
    public LinkedList<Integer> path = new LinkedList<Integer>();

    public List<List<Integer>> subsets(int[] nums) {
        res = new ArrayList<>();
        backtrace(nums, 0);
        return res;
    }

    public void backtrace(int[] nums, int index){
        if(index > nums.length) return;

        res.add(new ArrayList<>(){{addAll(path);}});

        for(int i = index; i < nums.length; i++){
            path.add(nums[i]);
            backtrace(nums, i + 1);
            path.removeLast();
        }
    }
}